### PR TITLE
[Relay][Training] Make AutoDiff thread through global function.

### PR DESCRIPTION
**Pull request by MarisaKirisame**
@junrushao1994 @jroesch @vinx13 @altanh @icemelon9 @hypercubestart @t-vi can you guys help review?

### CODE REVIEW

File: src/relay/transforms/gradient.cc [hunk 1]
```diff
@@ -559,6 +581,39 @@ struct ReverseAD : ExprMutator {
...
 
+  Expr VisitExpr_(const GlobalVarNode* op) final {
+    // todo: concatenating string to add attribute seems like a brittle hack.
+    // maybe get module indexed by a rose tree of string?
+    CHECK(mod.defined());
+    auto orig_gv = GetRef<GlobalVar>(op);
+    if (ad_gvars->count(orig_gv) == 0) {
+      GlobalVar gv(op->name_hint + "_grad");
+      (*ad_gvars)[orig_gv] = gv;
+      Function orig_f = Downcast<Function>(mod.value()->Lookup(GetRef<GlobalVar>(op)));
```
Comment (hypercubestart) [inline src/relay/transforms/gradient.cc:592]: mod.value()->Lookup(orig_gv)

Comment (hypercubestart) [inline src/relay/transforms/gradient.cc:592]: would it be a good idea to DeDup here?

File: tests/python/relay/test_pass_gradient.py [hunk 2]
```diff
@@ -341,5 +357,19 @@ def test_no_duplication():
...
+    t = relay.TensorType([])
+    x = relay.Var('x', t)
+    d = GlobalVar('double')
+    m[d] = relay.Function([x], x + x)
+    y = relay.Var('y', t)
+    q = GlobalVar('q')
+    m[q] = relay.Function([y], d(d(y)))
+    g = GlobalVar('grad')
+    m[g] = tvm.relay.transform.gradient(q, m)
+
```
Comment (hypercubestart) [inline tests/python/relay/test_pass_gradient.py:372]: add type and value check

File: src/relay/transforms/gradient.cc [hunk 3]
```diff
@@ -85,7 +85,7 @@ Expr DeGlobal(const Optional<IRModule>& mod, const Expr& e) {
   if (mod.defined() && x) {
     BaseFunc base_func = mod.value()->Lookup(GetRef<GlobalVar>(x));
     if (auto* n = base_func.as<FunctionNode>()) {
```
[the remark this answers is not in the document]
Comment (jroesch) [reply src/relay/transforms/gradient.cc:87]: It could point to a PrimFunction

File: src/relay/transforms/gradient.cc [hunk 4]
```diff
@@ -438,12 +449,17 @@ Expr BPEmpty() {
 
 struct ReverseAD : ExprMutator {
   using ADVarMap = std::unordered_map<Var, Var, ObjectPtrHash, ObjectPtrEqual>;
-
+  using ADGVarMap = std::unordered_map<GlobalVar, GlobalVar, ObjectPtrHash, ObjectPtrEqual>;
```
Comment (jroesch) [inline src/relay/transforms/gradient.cc:452]: Better name here

File: src/relay/transforms/gradient.cc [hunk 5]
```diff
@@ -438,12 +449,17 @@ Expr BPEmpty() {
...
+  Optional<IRModule> mod;
```
Comment (jroesch) [inline src/relay/transforms/gradient.cc:453]: Is this really optional?

File: src/relay/transforms/gradient.cc [hunk 6]
```diff
@@ -518,22 +533,29 @@ struct ReverseAD : ExprMutator {
...
+          tvm::Array<Expr> rev = rev_map[op_ref](orig, GetGrad(call->checked_type(), ret, ll));
+          CHECK(args.size() == rev.size());
+          for (size_t i = 0; i < args.size(); ++i) {
+            UpdateGrad(call->args[i]->checked_type(), args[i], rev[i], ll);
+          }
+          return Call(bpv, {});
+        });
+        Expr nbp = Function({}, nbp_body, TupleType::Empty(), {});
         ll->Push(RefWrite(bp, transform::ToANormalForm(nbp)));
         // TODO(@M.K.): ToANF should be called on rev. Enhance ToANF for that.
```
Comment (jroesch) [inline src/relay/transforms/gradient.cc:546]: Is this comment still valid? we should add to the TODO list if so

File: src/relay/transforms/gradient.cc (hunk 5 above)
Comment (MarisaKirisame) [reply src/relay/transforms/gradient.cc:453]: yes - sometime it is call on naked expr. i could refactor to require it to always be here, if this is what you want

File: src/relay/transforms/gradient.cc (hunk 6 above)
Comment (MarisaKirisame) [reply src/relay/transforms/gradient.cc:546]: yes, i just added it yesterday, what list

File: src/relay/transforms/gradient.cc (hunk 5 above)
Comment (jroesch) [reply src/relay/transforms/gradient.cc:453]: Can you just add a note about this to your on-going doc on things we might want to refactor?

Comment (MarisaKirisame) [reply src/relay/transforms/gradient.cc:453]: you mean the list from ~2 years ago? that is very out of dated. I think we can just grep -r for TODO(@M.K.) in the source.

**Pull request merged**